composer/core/state.py

```diff
@@ -1168,9 +1172,21 @@ def load_model_state(
                         log.debug(
                             f'Loading model state dict with strict={strict} and FSDP state_dict_type={self.fsdp_state_dict_type}'
                         )
-                        with fsdp_state_dict_type_context(self.model, state_dict_type=self.fsdp_state_dict_type):
-                            missing_keys, unexpected_keys = self.model.load_state_dict(state_dict['model'],
-                                                                                       strict=strict)
+                        # If the desired state_dict_type is sharded, but the checkpoint is monolithic,
+                        # we need to load the monolithic state dict and then scatter it to the sharded model
+                        try:
+                            with fsdp_state_dict_type_context(self.model, state_dict_type=self.fsdp_state_dict_type):
+                                missing_keys, unexpected_keys = self.model.load_state_dict(state_dict['model'],
+                                                                                           strict=strict)
+                        except AssertionError as e:
```

Comment:
where is this `AssertionError` raised?

Reply:
```
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /llm-foundry/scripts/train/train.py:611 in <module>                          │
│                                                                              │
│   608 │   cfg = om.merge(yaml_cfg, cli_cfg)                                  │
│   609 │   om.resolve(cfg)                                                    │
│   610 │   assert isinstance(cfg, DictConfig)                                 │
│ ❱ 611 │   main(cfg)                                                          │
│   612                                                                        │
│                                                                              │
│ /llm-foundry/scripts/train/train.py:543 in main                              │
│                                                                              │
│   540 │                                                                      │
│   541 │   # Build the Trainer                                                │
│   542 │   log.info('Building trainer...')                                    │
│ ❱ 543 │   trainer = Trainer(                                                 │
│   544 │   │   run_name=run_name,                                             │
│   545 │   │   seed=seed,                                                     │
│   546 │   │   model=model,                                                   │
│                                                                              │
│ /usr/lib/python3/dist-packages/composer/trainer/trainer.py:1495 in __init__  │
│                                                                              │
│   1492 │   │   │   │   if wandb.run is None:                                 │
│   1493 │   │   │   │   │   load_object_store.init(self.state, self.logger)   │
│   1494 │   │   │   _, _, parsed_load_path = parse_uri(load_path)             │
│ ❱ 1495 │   │   │   self._rng_state = checkpoint.load_checkpoint(             │
│   1496 │   │   │   │   state=self.state,                                     │
│   1497 │   │   │   │   logger=self.logger,                                   │
│   1498 │   │   │   │   path=parsed_load_path,                                │
│                                                                              │
│ /usr/lib/python3/dist-packages/composer/utils/checkpoint.py:402 in           │
│ load_checkpoint                                                              │
│                                                                              │
│    399 │   │   │   │   │   fsdp_sharded_state_dict_enabled=state.fsdp_sharde │
│    400 │   │   │   │   │   deepspeed_sharded_checkpoint=is_model_deepspeed(s │
│    401 │   │   │   │   )                                                     │
│ ❱  402 │   │   │   │   rng_state_dicts = _restore_checkpoint(                │
│    403 │   │   │   │   │   state,                                            │
│    404 │   │   │   │   │   logger,                                           │
│    405 │   │   │   │   │   composer_states_filepath,                         │
│                                                                              │
│ /usr/lib/python3/dist-packages/composer/utils/checkpoint.py:893 in           │
│ _restore_checkpoint                                                          │
│                                                                              │
│    890 │   │   │   algorithm_passes=algorithm_passes,                        │
│    891 │   │   )                                                             │
│    892 │   if not load_weights_only:                                         │
│ ❱  893 │   │   state.load_state_dict(                                        │
│    894 │   │   │   state_dict['state'],                                      │
│    895 │   │   │   logger,                                                   │
│    896 │   │   │   exclude_algorithms=exclude_algorithms,                    │
│                                                                              │
│ /usr/lib/python3/dist-packages/composer/core/state.py:1281 in                │
│ load_state_dict                                                              │
│                                                                              │
│   1278 │   │                                                                 │
│   1279 │   │   # Call load_model_state first since it applies required algor │
│   1280 │   │   if 'model' in state:                                          │
│ ❱ 1281 │   │   │   self.load_model_state(                                    │
│   1282 │   │   │   │   state,                                                │
│   1283 │   │   │   │   logger,                                               │
│   1284 │   │   │   │   strict=strict,                                        │
│                                                                              │
│ /usr/lib/python3/dist-packages/composer/core/state.py:1170 in                │
│ load_model_state                                                             │
│                                                                              │
│   1167 │   │   │   │   │   │   │   f'Loading model state dict with strict={s │
│   1168 │   │   │   │   │   │   )                                             │
│   1169 │   │   │   │   │   │   with fsdp_state_dict_type_context(self.model, │
│ ❱ 1170 │   │   │   │   │   │   │   missing_keys, unexpected_keys = self.mode │
│   1171 │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   │   │ │
│   1172 │   │   │   │   │   else:                                             │
│   1173 │   │   │   │   │   │   log.debug(f'Loading model state dict with str │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/nn/modules/module.py:2138 in            │
│ load_state_dict                                                              │
│                                                                              │
│   2135 │   │   │   │   │   "it should be done inplace."                      │
│   2136 │   │   │   │   )                                                     │
│   2137 │   │                                                                 │
│ ❱ 2138 │   │   load(self, state_dict)                                        │
│   2139 │   │   del load                                                      │
│   2140 │   │                                                                 │
│   2141 │   │   if strict:                                                    │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/nn/modules/module.py:2126 in load       │
│                                                                              │
│   2123 │   │   │   │   if child is not None:                                 │
│   2124 │   │   │   │   │   child_prefix = prefix + name + '.'                │
│   2125 │   │   │   │   │   child_state_dict = {k: v for k, v in local_state_ │
│ ❱ 2126 │   │   │   │   │   load(child, child_state_dict, child_prefix)       │
│   2127 │   │   │                                                             │
│   2128 │   │   │   # Note that the hook can modify missing_keys and unexpect │
│   2129 │   │   │   incompatible_keys = _IncompatibleKeys(missing_keys, unexp │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/nn/modules/module.py:2120 in load       │
│                                                                              │
│   2117 │   │   │   local_metadata = {} if metadata is None else metadata.get │
│   2118 │   │   │   if assign:                                                │
│   2119 │   │   │   │   local_metadata['assign_to_params_buffers'] = assign   │
│ ❱ 2120 │   │   │   module._load_from_state_dict(                             │
│   2121 │   │   │   │   local_state_dict, prefix, local_metadata, True, missi │
│   2122 │   │   │   for name, child in module._modules.items():               │
│   2123 │   │   │   │   if child is not None:                                 │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/nn/modules/module.py:1991 in            │
│ _load_from_state_dict                                                        │
│                                                                              │
│   1988 │   │   │   │   :meth:`~torch.nn.Module.load_state_dict`              │
│   1989 │   │   """                                                           │
│   1990 │   │   for hook in self._load_state_dict_pre_hooks.values():         │
│ ❱ 1991 │   │   │   hook(state_dict, prefix, local_metadata, strict, missing_ │
│   1992 │   │                                                                 │
│   1993 │   │   persistent_buffers = {k: v for k, v in self._buffers.items()  │
│   1994 │   │   local_name_params = itertools.chain(self._parameters.items(), │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/nn/modules/module.py:69 in __call__     │
│                                                                              │
│     66 │   │   │   module = self.module()                                    │
│     67 │   │   │   if module is None:                                        │
│     68 │   │   │   │   raise RuntimeError("You are trying to call the hook o │
│ ❱   69 │   │   │   return self.hook(module, *args, **kwargs)                 │
│     70 │   │   return self.hook(*args, **kwargs)                             │
│     71 │                                                                     │
│     72 │   def __getstate__(self) -> Dict:                                   │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/utils/_contextlib.py:115 in             │
│ decorate_context                                                             │
│                                                                              │
│   112 │   @functools.wraps(func)                                             │
│   113 │   def decorate_context(*args, **kwargs):                             │
│   114 │   │   with ctx_factory():                                            │
│ ❱ 115 │   │   │   return func(*args, **kwargs)                               │
│   116 │                                                                      │
│   117 │   return decorate_context                                            │
│   118                                                                        │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_state_dict_utils.py:8 │
│ 02 in _pre_load_state_dict_hook                                              │
│                                                                              │
│   799 │   │   if fsdp_state._device_handle.is_available():                   │
│   800 │   │   │   fsdp_state._device_handle.synchronize()                    │
│   801 │   │   # Dispatch into state_dict specific implementation of pre-hook │
│ ❱ 802 │   │   _pre_load_state_dict_hook_fn[fsdp_state._state_dict_type](     │
│   803 │   │   │   module, fsdp_state, state_dict, prefix                     │
│   804 │   │   )                                                              │
│   805                                                                        │
│                                                                              │
│ /usr/lib/python3/dist-packages/composer/trainer/mosaic_fsdp_utils.py:574 in  │
│ _sharded_pre_load_state_dict_hook                                            │
│                                                                              │
│    571 │   │                                                                 │
│    572 │   │   if not fsdp_state._state_dict_config.use_dtensor:             │
│    573 │   │   │   # All-gather the param (ShardedTensor)                    │
│ ❱  574 │   │   │   param, shards = _ext_pre_load_state_dict_transform(param) │
│    575 │   │   │                                                             │
│    576 │   │   │   assert len(shards) < 2, ('Expects 0 or 1 shard per rank ' │
│    577 │   │   │   │   │   │   │   │   │    f'but got {len(shards)} shards o │
│                                                                              │
│ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_fsdp_extensions.py:13 │
│ 3 in _ext_pre_load_state_dict_transform                                      │
│                                                                              │
│   130 │   if _extensions is not None:                                        │
│   131 │   │   return _extensions.pre_load_state_dict_transform(tensor)       │
│   132 │                                                                      │
│ ❱ 133 │   assert type(tensor) is ShardedTensor                               │
│   134 │   shards = tensor.local_shards()                                     │
│   135 │   return (tensor, shards)                                            │
│   136                                                                        │
╰──────────────────────────────────────────────────────────────────────────────╯
AssertionError
```